lib/model/model.go

```diff
@@ -1984,98 +2003,43 @@ func (m *Model) internalScanFolderSubdirs(ctx context.Context, folder string, su
 		}
 	}()
 
-	for f := range fchan {
-		if len(batch) == maxBatchSizeFiles || batchSizeBytes > maxBatchSizeBytes {
-			if err := runner.CheckHealth(); err != nil {
+	delDirStack := make([]protocol.FileInfo, 0)
+	for r := range rchan {
+		if err := checkBatch(); err != nil {
+			l.Debugln("Stopping scan of folder %s due to: %s", folderCfg.Description(), err)
+			return err
+		}
+
+		for len(delDirStack) != 0 && !strings.HasPrefix(r.New.Name, delDirStack[len(delDirStack)-1].Name+string(fs.PathSeparator)) {
```

Comment:
Comment on what's being accomplished here would be appreciated. It's not clear to me at first glance. Also the `delDirStack[len(delDirStack)-1]` thing could have a name as it's repeated a lot.